CleanStackTraceExtension does not clear related exceptions

**The problem**

CleanStackTraceExtension does not clear related exceptions. Add boolean option `related`.

Example:

```java
step("step", () -> {
  Exception mainException = new Exception();
  mainException.initCause(new Exception());
  mainException.addSuppressed(new Exception());
  throw mainException;
});
```

This code will produce:

```
java.lang.Exception
	at my.proj.Tests.lambda$test$0(Tests.java:35)
	at my.proj.Tests.test(Tests.java:34)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	Suppressed: java.lang.Exception
		at my.proj.Tests.lambda$test$0(Tests.java:37)
		at org.stebz.executor.StepExecutor$Of.lambda$execute$11(StepExecutor.java:178)
		at org.stebz.executor.StepExecutor$Of.exec(StepExecutor.java:235)
		at org.stebz.executor.StepExecutor$Of.execute(StepExecutor.java:177)
		at org.stebz.StebzMethods.step(StebzMethods.java:490)
		... 4 more
Caused by: java.lang.Exception
	at my.proj.Tests.lambda$test$0(Tests.java:36)
	at org.stebz.executor.StepExecutor$Of.lambda$execute$11(StepExecutor.java:178)
	at org.stebz.executor.StepExecutor$Of.exec(StepExecutor.java:235)
	at org.stebz.executor.StepExecutor$Of.execute(StepExecutor.java:177)
	at org.stebz.StebzMethods.step(StebzMethods.java:490)
	... 4 more
```

**Environment**

- Stebz version 1.11

--- a/stebz-clean-stack-trace/src/main/java/org/stebz/extension/CleanStackTraceExtension.java
+++ b/stebz-clean-stack-trace/src/main/java/org/stebz/extension/CleanStackTraceExtension.java
@@ -29,6 +29,8 @@ import org.stebz.util.container.NullableOptional;
 import org.stebz.util.property.PropertiesReader;
 
 import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
 import java.util.function.Predicate;
 
 /**
@@ -42,6 +44,7 @@ public class CleanStackTraceExtension implements BeforeStepFailure {
   private final Predicate<StackTraceElement> stackTraceElementFilter;
   private final boolean enabled;
   private final int order;
+  private final boolean clearRelated;
 
   /**
    * Ctor.
@@ -58,21 +61,22 @@ public class CleanStackTraceExtension implements BeforeStepFailure {
   public CleanStackTraceExtension(final PropertiesReader properties) {
     this.enabled = properties.getBoolean("stebz.extensions.cleanStackTrace.enabled", true);
     this.order = properties.getInteger("stebz.extensions.cleanStackTrace.order", MIDDLE_ORDER);
-    final boolean stebzLines = properties.getBoolean("stebz.extensions.cleanStackTrace.stebzLines", true);
-    Boolean aspectjLines = properties.getBoolean("stebz.extensions.cleanStackTrace.aspectjLines", null);
-    if (aspectjLines == null) {
-      aspectjLines = isStebzAnnotationsUsed();
+    this.clearRelated = properties.getBoolean("stebz.extensions.cleanStackTrace.clearRelated", true);
+    final boolean clearStebzLines = properties.getBoolean("stebz.extensions.cleanStackTrace.clearStebzLines", true);
+    Boolean clearAspectjLines = properties.getBoolean("stebz.extensions.cleanStackTrace.clearAspectjLines", null);
+    if (clearAspectjLines == null) {
+      clearAspectjLines = isStebzAnnotationsUsed();
     }
-    if (stebzLines && aspectjLines) {
+    if (clearStebzLines && clearAspectjLines) {
       this.stackTraceElementFilter = stElem ->
         !stElem.getClassName().startsWith(STEBZ_CLASS_NAME_PREFIX)
           && !stElem.getClassName().startsWith(ASPECTJ_CLASS_NAME_PREFIX)
           && !stElem.getClassName().contains(ASPECTJ_CLASS_NAME_PART)
           && !stElem.getMethodName().contains(ASPECTJ_METHOD_NAME_PART);
-    } else if (stebzLines) {
+    } else if (clearStebzLines) {
       this.stackTraceElementFilter = stElem ->
         !stElem.getClassName().startsWith(STEBZ_CLASS_NAME_PREFIX);
-    } else if (aspectjLines) {
+    } else if (clearAspectjLines) {
       this.stackTraceElementFilter = stElem ->
         !stElem.getClassName().startsWith(ASPECTJ_CLASS_NAME_PREFIX)
           && !stElem.getClassName().contains(ASPECTJ_CLASS_NAME_PART)
@@ -100,9 +104,16 @@ public class CleanStackTraceExtension implements BeforeStepFailure {
   public void beforeStepFailure(final StepObj<?> step,
                                 final NullableOptional<Object> context,
                                 final Throwable exception) {
-    if (!this.enabled) {
-      return;
+    if (this.enabled) {
+      if (this.clearRelated) {
+        relatedExceptions(new HashSet<>(), exception).forEach(this::clearException);
+      } else {
+        this.clearException(exception);
+      }
     }
+  }
+
+  private void clearException(final Throwable exception) {
     final StackTraceElement[] originST = exception.getStackTrace();
     if (originST.length != 0) {
       exception.setStackTrace(
@@ -111,5 +122,17 @@ public class CleanStackTraceExtension implements BeforeStepFailure {
           .toArray(StackTraceElement[]::new)
       );
     }
+  }
+
+  private static Set<Throwable> relatedExceptions(final Set<Throwable> exceptions,
+                                                  final Throwable mainEx) {
+    for (Throwable currentEx = mainEx; currentEx != null; currentEx = currentEx.getCause()) {
+      if (exceptions.add(currentEx)) {
+        for (final Throwable suppressedEx : currentEx.getSuppressed()) {
+          relatedExceptions(exceptions, suppressedEx);
+        }
+      }
+    }
+    return exceptions;
   }
 }

--- a/stebz-clean-stack-trace/src/test/java/org/stebz/extension/CleanStackTraceExtensionTest.java
+++ b/stebz-clean-stack-trace/src/test/java/org/stebz/extension/CleanStackTraceExtensionTest.java
@@ -40,11 +40,22 @@ final class CleanStackTraceExtensionTest {
 
   @Test
   void extensionShouldRemoveStebzLines() {
-    final Throwable exception = new Throwable();
+    final Throwable causeEx = new Throwable();
+    final Throwable suppressedEx = new Throwable();
+    final Throwable causeSuppressedEx = new Throwable();
+    final Throwable mainException = new Throwable(causeEx);
+    causeEx.addSuppressed(causeSuppressedEx);
+    mainException.addSuppressed(suppressedEx);
     final CleanStackTraceExtension extension = new CleanStackTraceExtension(new PropertiesReader.Of(new Properties()));
 
-    extension.beforeStepFailure(mock(StepObj.class), NullableOptional.empty(), exception);
-    assertThat(exception.getStackTrace())
+    extension.beforeStepFailure(mock(StepObj.class), NullableOptional.empty(), mainException);
+    assertThat(mainException.getStackTrace())
+      .allMatch(stElem -> !stElem.getClassName().startsWith("org.stebz."));
+    assertThat(causeEx.getStackTrace())
+      .allMatch(stElem -> !stElem.getClassName().startsWith("org.stebz."));
+    assertThat(suppressedEx.getStackTrace())
+      .allMatch(stElem -> !stElem.getClassName().startsWith("org.stebz."));
+    assertThat(causeSuppressedEx.getStackTrace())
       .allMatch(stElem -> !stElem.getClassName().startsWith("org.stebz."));
   }
 }

